game/src/edit/roads.rs

```diff
@@ -1,84 +1,461 @@
-use map_gui::tools::PromptInput;
-use map_model::{Direction, LaneSpec, LaneType, Road};
-use widgetry::{EventCtx, State};
+use geom::Distance;
+use map_gui::render::{Renderable, OUTLINE_THICKNESS};
+use map_model::{
+    Direction, EditCmd, EditRoad, LaneID, LaneSpec, LaneType, Road, RoadID, NORMAL_LANE_THICKNESS,
+};
+use widgetry::{
+    lctrl, Choice, Color, Drawable, EventCtx, GeomBatch, GfxCtx, HorizontalAlignment, Key, Line,
+    Outcome, Panel, State, Text, TextExt, VerticalAlignment, Widget,
+};
 
 use crate::app::{App, Transition};
-use crate::edit::apply_map_edits;
-
-/// Specify the lane types for a road using a text box. This is a temporary UI to start
-/// experimenting with widening roads. It'll be replaced by a real UI once the design is ready.
-pub fn prompt_for_lanes(ctx: &mut EventCtx, road: &Road) -> Box<dyn State<App>> {
-    let r = road.id;
-    PromptInput::new(
-        ctx,
-        "Define lanes_ltr",
-        lanes_to_string(road),
-        Box::new(move |string, ctx, app| {
-            // We're selecting a lane before this, but the ID is probably about to be invalidated.
-            app.primary.current_selection = None;
-
-            let mut edits = app.primary.map.get_edits().clone();
-            edits.commands.push(app.primary.map.edit_road_cmd(r, |new| {
-                new.lanes_ltr = string_to_lanes(string.clone());
-            }));
-            apply_map_edits(ctx, app, edits);
-            Transition::Multi(vec![Transition::Pop, Transition::Pop])
-        }),
-    )
+use crate::edit::{apply_map_edits, speed_limit_choices};
+
+pub struct RoadEditor {
+    r: RoadID,
+    current_lane: Option<LaneID>,
+    top_panel: Panel,
```

Comment:
Seeing it, in action, it feels like the submit/undo/redo buttons on the top panel would make more sense directly on the panel where you'd doing all the editing, like any other form with fields and a submit button.

Is there a rationale for them being in a separate top panel?  This isn't a blocking concern, but wanted to mention it as you continue to iterate.

Reply:
Mostly I'm blindly following the design:
![Screenshot from 2021-04-22 07-42-42](https://user-images.githubusercontent.com/1664407/115734180-5ab45680-a33e-11eb-9123-8c930929568d.png)
When the top panel has "cancel" and "edit multiple", I think undo/redo makes more sense. It's similar to the signal editor
![Screenshot from 2021-04-22 07-44-07](https://user-images.githubusercontent.com/1664407/115734422-8b948b80-a33e-11eb-9fb6-f87b97cc1fa8.png)
Where we group "meta" actions for the process of editing separately